Transient 400: Bad Request responses

jetty-9.4.45.v20220128 Compact 3 (Not yet released)

java -version
openjdk version "1.8.0_322"
OpenJDK Runtime Environment (Zulu 8.60.0.22-SA-qnx_aarch32hf) (build 1.8.0_322-b06, profile compact3)
OpenJDK Client VM (Zulu 8.60.0.22-SA-qnx_aarch32hf) (build 25.322-b06, mixed mode)

QNX 7.0 ARMLEv7

I understand that this version is not yet released, but I have done some early testing against the 9.4.45.v20220128 tag made yesterday in anticipation of adopting the release as soon as it is made available. I am seeing a substantial regression in our application behavior where a 400 Bad Request response is being returned intermittently. The only change in our runtime environment is that the 9.4.45 Jetty version (jetty-all-compact3-9.4.45.v20220128.jar) replaces the 9.4.44 Jetty version (jetty-all-compact3-9.4.44.v20210927.jar) in our classpath. Sometimes the request goes through normally (200 response), and other times the page fails to load, or partially loads (with 400 response). The page always loads normally with 9.4.44.

The request in question is HTTP so no HTTPS/TLS negotiation is a factor.

I apologize that this description is vague, and that this 9.4.45 version might not be released yet if still a work-in-progress, but I wanted to provide early feedback to see if this change in behavior rings any bells about what may have changed between 9.4.44 and 9.4.45.20220128 that might bring about this behavior (and prevent our uptake of 9.4.45). Is there any particular release note I should be paying attention to that might require change in our application configuration? We've been using the 9.4.X branch since 9.4.26 and have not seen this type of behavior until testing with recent 9.4.45.

Thanks, please let me know if you have any questions that might help me clarify this behavior.

Recently upgraded to jetty-9.4.45.v20220128 to get the ALPN fix, but believe I am running into this issue. This does not happen in 9.4.44.

```2022-02-11 09:12:13,507,DEBUG,org.eclipse.jetty.http.HttpParser - [qtp848102479-250]  [::] Parse exception: HttpParser{s=HEADER,0 of -1} for HttpChannelOverHttp@2bef0009{s=HttpChannelState@46d6747c{s=IDLE rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=0,c=false/false,a=IDLE,uri=null,age=0}
org.eclipse.jetty.http.BadMessageException: 400: Bad Request
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1640) [jetty-http-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.server.HttpConnection.parseRequestBuffer(HttpConnection.java:381) [jetty-server-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:265) [jetty-server-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [jetty-io-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) [jetty-io-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:555) [jetty-io-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:410) [jetty-io-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:164) [jetty-io-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) [jetty-io-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) [jetty-io-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) [jetty-util-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) [jetty-util-9.4.45.v20220203.jar:9.4.45.v20220203]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_312]
Caused by: java.lang.IllegalArgumentException: Capacity 131072 > 65534
	at org.eclipse.jetty.util.ArrayTernaryTrie.<init>(ArrayTernaryTrie.java:151) ~[jetty-util-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.util.ArrayTernaryTrie.<init>(ArrayTernaryTrie.java:133) ~[jetty-util-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.http.HttpParser.parsedHeader(HttpParser.java:1077) ~[jetty-http-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.http.HttpParser.parseFields(HttpParser.java:1258) ~[jetty-http-9.4.45.v20220203.jar:9.4.45.v20220203]
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1528) [jetty-http-9.4.45.v20220203.jar:9.4.45.v20220203]
	... 12 more```

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java
@@ -27,6 +27,7 @@ import java.util.concurrent.CopyOnWriteArrayList;
 import org.eclipse.jetty.http.CookieCompliance;
 import org.eclipse.jetty.http.HttpMethod;
 import org.eclipse.jetty.http.HttpScheme;
+import org.eclipse.jetty.util.ArrayTernaryTrie;
 import org.eclipse.jetty.util.HostPort;
 import org.eclipse.jetty.util.Jetty;
 import org.eclipse.jetty.util.TreeTrie;
@@ -434,6 +435,10 @@ public class HttpConfiguration implements Dumpable
      */
     public void setHeaderCacheSize(int headerCacheSize)
     {
+        if (headerCacheSize > ArrayTernaryTrie.MAX_CAPACITY)
+        {
+            throw new IllegalArgumentException("headerCacheSize " + headerCacheSize + " exceeds maximum value of " + ArrayTernaryTrie.MAX_CAPACITY);
+        }
         _headerCacheSize = headerCacheSize;
     }
 

--- a/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java
+++ b/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java
@@ -74,7 +74,7 @@ public class ArrayTernaryTrie<V> extends AbstractTrie<V>
      * the 16 bit indexes can overflow and the trie
      * cannot find existing entries anymore.
      */
-    private static final int MAX_CAPACITY = Character.MAX_VALUE - 1;
+    public static final int MAX_CAPACITY = Character.MAX_VALUE - 1;
 
     /**
      * The Trie rows in a single array which allows a lookup of row,character

